Sources/ClientRuntime/Networking/Http/AsyncThrowingStreamReader.swift

```diff
@@ -0,0 +1,106 @@
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
+import AwsCommonRuntimeKit
+
+/// A `StreamReader` that reads `AsyncThrowingStream<Data, Error>`
+/// Event Stream messages are sent as a series of `Data` chunks
+public class AsyncThrowingStreamReader: StreamReader {
+    public var availableForRead: UInt {
+        get {
+            fatalError("availableForRead is not supported with async streams")
+        }
+        set {
+            fatalError("availableForRead is not supported with async streams")
+        }
+    }
+
+    private var _hasFinishedWriting: Bool
+    public var hasFinishedWriting: Bool {
+        get {
+            lock.withLockingClosure {
+                return _hasFinishedWriting
+            }
+        }
+        set {
+            lock.withLockingClosure {
+                _hasFinishedWriting = newValue
+            }
+        }
+    }
+
+    internal let stream: AsyncThrowingStream<Data, Error>
+    internal var buffer: Data = Data()
+    private let lock = NSLock()
+
+    public init(stream: AsyncThrowingStream<Data, Error>) {
+        self.stream = stream
+        self._hasFinishedWriting = false
+    }
+
+    public func read(maxBytes: UInt?, rewind: Bool = false) -> AwsCommonRuntimeKit.ByteBuffer {
+        if rewind {
+            fatalError("Rewind is not supported with async streams")
+        }
+
+        return lock.withLockingClosure {
+            return _unsafeWait {
+                return await self.read(maxBytes: maxBytes)
+            }
+        }
+    }
+
+    private func read(maxBytes: UInt?) async -> AwsCommonRuntimeKit.ByteBuffer {
+        var data = Data()
+        var remaining = maxBytes ?? UInt.max
+
+        // copy anything leftover from the last read
+        if !buffer.isEmpty {
+            let bytesToCopy = Swift.min(remaining, UInt(buffer.count))
+            data.append(buffer[0..<Int(bytesToCopy)])
+            remaining -= bytesToCopy
+            buffer = buffer[Int(bytesToCopy)..<buffer.count]
+        }
+
+        var iterator = stream.makeAsyncIterator()
+        while remaining > 0 {
+            do {
+                let next = try await iterator.next()
+                if let next: Data = next {
+                    // copy remaining bytes from the next chunk
+                    let bytesToCopy = Swift.min(remaining, UInt(next.count))
+                    data.append(next[0..<Int(bytesToCopy)])
+                    remaining -= bytesToCopy
+
+                    // store leftover bytes for next read
+                    if next.count > Int(remaining) {
+                        buffer = Data(next[Int(bytesToCopy)..<next.count])
+                    }
+                } else {
+                    break
+                }
+            } catch {
+                // TODO: handle error
+                fatalError("TODO: handle error \(error)")
+            }
+        }
+        return ByteBuffer(data: data)
+    }
+
+    public func seek(offset: Int) throws {
```

Comment:
The fact that so many methods from the `StreamReader` protocol fail if called suggests that this protocol is not the right abstraction for an asynchronous stream, or that the protocol needs to be broken down into narrower sub-protocols that certain sub-types of streams may adopt.

Reply:
I agree with Josh here, although this might be out of scope for this PR and this feature, but we should create an issue to fix this up.